packages/fxa-customs-server/lib/dataflow.js

```diff
@@ -54,24 +87,154 @@ module.exports = function (config, log) {
     })
 
     messageCount++
+
+    _processMessage(message)
   }
 
-  const messageHandler = message => {
-    logMessage(message)
+  /**
+   * Parse a message and check the metadata for any unexpected rate limits
+   *
+   * @param {Object} message
+   * @param {Function} [_parseMessage=parseMessage]
+   * @param {Function} [_checkForUnexpectedDataflowBlock=checkForUnexpectedDataflowBlock]
+   */
+  function processMessage(message, _parseMessage=parseMessage, _checkForUnexpectedDataflowBlock=checkForUnexpectedDataflowBlock) {
+    let dataflowBlock
 
-    message.ack()
+    try {
+      dataflowBlock = _parseMessage(message)
+    } catch (err) {
+      log.error({
+        op: 'dataflow.message.invalid',
+        reason: err.message
+      })
+    }
+
+    if (dataflowBlock) {
+      _checkForUnexpectedDataflowBlock(dataflowBlock)
+    }
   }
 
-  subscription.on('message', messageHandler)
+  /**
+   * Parse a message that arrives from PubSub, returning
+   * the information about the DataFlow block.
+   *
+   * @param {Object} message
+   * @returns {Object}
+   */
+  function parseMessage (message) {
+    if (! message) {
+      throw new Error('missing message')
+    }
+    if (! message.data) {
+      throw new Error('missing message data')
+    }
+    if (! Buffer.isBuffer(message.data)) {
+      throw new Error('message data is not a Buffer')
+    }
 
-  const logError = error => {
-    // logs errors but does not cause the server
-    // to shut down.
-    log.error({
-      op: 'fxa.customs.dataflow.error',
-      error: String(error)
+  // message.data is a Buffer
+    const body = message.data.toString()
+    const parsedBody = JSON.parse(body)
+
+    if (! parsedBody.metadata) {
+      throw new Error('missing metadata')
+    }
+
+    if (! Array.isArray(parsedBody.metadata)) {
+      throw new Error('metadata is not an Array')
+    }
+
+    const dataflowBlock = {}
+
+    parsedBody.metadata.forEach(({ key, value }) => {
+      if (typeof key === 'undefined') {
+        throw new Error('missing metadata key')
+      }
+      if (typeof value === 'undefined') {
+        throw new Error('missing metadata value')
+      }
+      dataflowBlock[key] = value
     })
+
+    return dataflowBlock
+  }
+
+
+  /**
+   * Check for unexpected DataFlow blocks.
+   * Logs whenever an unexpected block is found.
+   *
+   * @param {Object} dataflowBlock
+   */
+  async function checkForUnexpectedDataflowBlock(dataflowBlock) {
```

Comment:
nit: Are we expecting other types of blocks in the `dataflow.js` file? If no, I think we could drop `Dataflow` in the function name here.